src/qt/optionsmodel.cpp

```diff
@@ -19,18 +19,59 @@
 #include <txdb.h>       // for -dbcache defaults
 #include <util/string.h>
 #include <validation.h> // For DEFAULT_SCRIPTCHECK_THREADS
+#include <wallet/wallet.h> // For DEFAULT_SPEND_ZEROCONF_CHANGE
 
 #include <QDebug>
 #include <QLatin1Char>
 #include <QSettings>
 #include <QStringList>
 
+#include <univalue.h>
+
 const char *DEFAULT_GUI_PROXY_HOST = "127.0.0.1";
 
 static const QString GetDefaultProxyAddress();
 
-OptionsModel::OptionsModel(QObject *parent, bool resetSettings) :
-    QAbstractListModel(parent)
+//! Convert enabled/size values to bitcoin -prune setting.
+static util::SettingsValue PruneSetting(bool prune_enabled, int prune_size_gb)
+{
+    assert(!prune_enabled || prune_size_gb >= 1); // PruneSizeGB and ParsePruneSizeGB never return less
+    return prune_enabled ? PruneGBtoMiB(prune_size_gb) : 0;
+}
+
+//! Get pruning enabled value to show in GUI from bitcoin -prune setting.
+static bool PruneEnabled(const util::SettingsValue& prune_setting)
+{
+    // -prune=1 setting is manual pruning mode, so disabled for purposes of the gui
+    return SettingToInt(prune_setting, 0) > 1;
+}
+
+//! Get pruning size value to show in GUI from bitcoin -prune setting. If
+//! pruning is not enabled, just show default recommended pruning size (2GB).
+static int PruneSizeGB(const util::SettingsValue& prune_setting)
+{
+    int value = SettingToInt(prune_setting, 0);
+    return value > 1 ? PruneMiBtoGB(value) : DEFAULT_PRUNE_TARGET_GB;
+}
+
+//! Parse pruning size value provided by user in GUI or loaded from a legacy
+//! QSettings source (windows registry key or qt .conf file). Smallest value
+//! that the GUI can display is 1 GB, so round up if anything less is parsed.
+static int ParsePruneSizeGB(const QVariant& prune_size)
+{
+    return std::max(1, prune_size.toInt());
+}
+
+struct ProxySetting {
+    bool is_set;
```

Comment:
_(continuing the proxy [discussion](https://github.com/bitcoin/bitcoin/pull/15936#pullrequestreview-937685759) here so that it is in its own thread, not in the main PR comments and so that it can be closed when resolved eventually)_

This applies to the options that have a user-input value and an additional checkbox to disable them: `prune`, `proxy` and `onion`. Naturally there are two things per option - the value typed by the user and the boolean checkbox (denoting enabled/disabled).

This PR uses a dummy value in `settings.json` to denote the "disabled" state: `0` for `prune` and `""` for `proxy`. The problem with this approach is that this loses the value typed by the user. This is surprising and annoying (having to re-enter it after temporary disabling it). Also, it is not consistent behavior:

* closing the config window and re-opening it => the values are still there, greyed out
* restarting `bitcoin-qt` => the values are deleted

Firefox for example has a proxy-config and it acts in a consistent way - the values, when disabled, are preserved (greyed out) even after restart. IMO this is [less surprising](https://en.wikipedia.org/wiki/Principle_of_least_astonishment).

I would suggest to save the "use prune", "use proxy" and "use proxy for tor" as explicit boolean entities in `settings.json` and preserve the user-entered values in `prune`, `proxy` and `onion`.

Actually, for `prune` the behavior is even a bit more confusing:
* enter 5 (some arbitrary value)
* disable the checkbox (5 remains in the box, greyed out)
* restart
* enable the checkbox
* surprise! the value is not 5, but 2

Reply:
> I'll try to implement a patch which implements the previous behavior you seem to like.

Hmm, I shouldn't be forcing my view here. I could be nuts and totally disconnected from realit:space_invader: Maybe at this point it makes sense to poke other reviewers for opinion. Anyone?

> I think it's ok to clear pruning or proxy values after a users disables pruning or proxying and then restarts.

I think it would be less surprising and more obvious what is going on if we clear the values after the user disables the option. Like, immediately after the click on the checkbox. Then the behavior will be consistent between "close & re-open the config dialog box" and "close & re-open the app".

To summarize, there are 3 approaches for when an option is disabled (sorted in order of my preference):

1. Preserve the value (greyed out) through config dialog re-open and through app restart. To achieve this, store an explicit `use_proxy: false` in `settings.json` and keep the proxy address in `proxy: "..."`. This means that a config which `bitcoind` does not recognize will appear in `settings.json` which will require a tweak here: https://github.com/bitcoin/bitcoin/blob/b297b945f7610772434817181ad12067b2832565/src/util/system.cpp#L569

2. Delete the value immediately after the checkbox is unchecked. This will make it obvious that the value is gone.

3. Preserve the value (greyed out) through config dialog re-open, but delete it through an app restart. This is how this PR works as of now (99b4ce266f62afc8a37575527d82ac6048f50f75).